### ISSUE TITLE
phpredis 4.2.0 segfaults with globals and shutdown functions with Redis objects

### ISSUE THREAD

**Issue by vityank** [phpredis/phpredis#1484]
### Expected behaviour
Script finishing after performing the redis operation.

### Actual behaviour
Segmentation fault then php-redis attempts to close persistent connection.

# Description
Seems that phpredis attempts to to close the persistent socket handle(???) during calls to regular redis commands(hIncrBy,hSet,etc).
Previous stable v4.1.1 works fine(It didn't closed persistent handles on disconnects).
The problem occurs then the following are met:
(a) There is a shutdown function registered which creates new instances of Redis.
(b) There is also some static/global variable holding Redis instance present.

### I'm seeing this behavior on
- OS: 
CentOS 7.5.1804
- Redis:
4.0.11
- PHP:
7.1.25
- SAPI:
FPM
- phpredis:
4.2.0

### Steps to reproduce, backtrace or example script

Example script:
```php
<?php

class RedisCheck_756
{
    public $redis;
    function __construct()
    {
        $this->redis = new Redis();
        $ret = $this->redis->pconnect('localhost', 6379, 30);
        if (!$ret)
            $this->redis = null;
    }

    public function testSet()
    {
        if (!$this->redis)
            return false;
        $tVal = 'myTestVal'.microtime();
        $res = $this->redis->set('testKey', $tVal);
        if (!$res)
            return false;
        $res = $this->redis->get('testKey');
        return ($res === $tVal);
    }

    function __destruct()
    {
        if ($this->redis)
        {
            $this->redis->close();
            $this->redis = null;
        }
    }
}

echo "Start test\n";

function runMe()
{
    $clTst = new RedisCheck_756();
    $r = $clTst->testSet();
    echo "Res: ".($r ? "OK" : "FAIL")."\n";
}

$alsoAddSomeGlobal = new RedisCheck_756();

register_shutdown_function(function ($dummy) {
    runMe();
}, null);

runMe(); // This is good
echo "Finishing test\n";
```

BT:

Program received signal SIGSEGV, Segmentation fault.
0x00000000005b9b27 in _php_stream_free (stream=0xffcd50, close_options=<optimized out>)
    at /root/basesys-components/php-7.1.25/main/streams/streams.c:372
372                     context = PHP_STREAM_CONTEXT(stream);
(gdb) bt
#0  0x00000000005b9b27 in _php_stream_free (stream=0xffcd50,
    close_options=<optimized out>)
    at /root/basesys-components/php-7.1.25/main/streams/streams.c:372
#1  0x00007f752d11909a in redis_sock_disconnect (
    redis_sock=redis_sock@entry=0x7f7540673300, force=force@entry=1)
    at /root/phpredis/library.c:1865
#2  0x00007f752d11a0c1 in redis_check_eof (redis_sock=redis_sock@entry=0x7f7540673300,
    no_throw=no_throw@entry=0) at /root/phpredis/library.c:190
#3  0x00007f752d11a3b4 in redis_sock_write (redis_sock=redis_sock@entry=0x7f7540673300,
    cmd=0x7f7540674e00 "*HIDDEN", sz=149) at /root/phpredis/library.c:2085
#4  0x00007f752d105f52 in zim_Redis_hIncrBy (execute_data=0x7f7540614860,
    return_value=0x7f7540614820) at /root/phpredis/redis.c:2209
#5  0x0000000000698558 in ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER ()
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:1097
#6  0x000000000064771b in execute_ex (ex=<optimized out>)
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:429
#7  0x000000000069a7e4 in zend_execute (op_array=0x7f754065d1c0,
    op_array@entry=0x7f7524a69130, return_value=return_value@entry=0x7f7540614770)
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:474
#8  0x0000000000602464 in zend_execute_scripts (type=type@entry=8,
    retval=0x7f7540614770, retval@entry=0x0, file_count=file_count@entry=3)
    at /root/basesys-components/php-7.1.25/Zend/zend.c:1482
#9  0x00000000005a3200 in php_execute_script (
    primary_file=primary_file@entry=0x7ffe3c6b9840)
    at /root/basesys-components/php-7.1.25/main/main.c:2577
#10 0x000000000042849c in main (argc=<optimized out>, argv=<optimized out>)
    at /root/basesys-components/php-7.1.25/sapi/fpm/fpm/fpm_main.c:1966

### I've checked
- [v] There is no similar issue from other users
- [v] Issue isn't fixed in `develop` branch

**Comment by yatsukhnenko** [comment]
I can't replicate this issue

```
php -v
PHP 7.1.25 (cli) (built: Dec 20 2018 16:08:14) ( NTS DEBUG )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.1.25, Copyright (c) 1999-2018, by Zend Technologies
```
```
php -d extension=modules/redis.so issue-1484.php 
Start test
Res: OK
Finishing test
Res: OK
```
Could you run your test script from cli?

**Comment by vityank** [comment]
Through i tested the case with cli as well before, here we go(Still considering other factors like glibc version and others):

/opt/php71/bin/php -d extension=redis.so /root/phpredis/testCloseOnShutdown.php
Start test
Res: OK
Finishing test
Res: OK
Segmentation fault

(gdb) bt
#0  0x00000000005b7ee4 in _php_stream_free (stream=<optimized out>, close_options=19)
    at /root/basesys-components/php-7.1.25/main/streams/streams.c:467
#1  0x00007fffdfd1909a in redis_sock_disconnect (redis_sock=0x7ffff32733c0,
    force=force@entry=1) at /root/phpredis/library.c:1865
#2  0x00007fffdfcf4906 in zim_Redis_close (execute_data=<optimized out>,
    return_value=0x7fffffffc950) at /root/phpredis/redis.c:1015
#3  0x0000000000696d76 in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER ()
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:970
#4  0x0000000000645a0b in execute_ex (ex=<optimized out>)
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:429
#5  0x00000000005f11c1 in zend_call_function (fci=fci@entry=0x7fffffffcb20,
    fci_cache=fci_cache@entry=0x7fffffffcaf0)
    at /root/basesys-components/php-7.1.25/Zend/zend_execute_API.c:855
#6  0x000000000061d220 in zend_call_method (object=object@entry=0x7fffffffcbd0,
    obj_ce=0x7ffff3204018, fn_proxy=fn_proxy@entry=0x7fffffffcbc8,
    function_name=function_name@entry=0xa5a5c1 "__destruct",
    function_name_len=function_name_len@entry=10, retval_ptr=retval_ptr@entry=0x0,
    param_count=param_count@entry=0, arg1=arg1@entry=0x0, arg2=arg2@entry=0x0)
    at /root/basesys-components/php-7.1.25/Zend/zend_interfaces.c:99
#7  0x0000000000636721 in zend_objects_destroy_object (object=<optimized out>)
    at /root/basesys-components/php-7.1.25/Zend/zend_objects.c:156
#8  0x000000000063b6bb in zend_objects_store_del (object=0x7ffff32571c0)
    at /root/basesys-components/php-7.1.25/Zend/zend_objects_API.c:162
#9  0x0000000000611ec5 in _zend_hash_del_el_ex (prev=<optimized out>, p=<optimized out>,
    idx=<optimized out>, ht=<optimized out>)
    at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:997
#10 _zend_hash_del_el (p=0x7ffff32581e0, idx=7, ht=0xdba9f0 <executor_globals+304>)
    at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:1020
#11 zend_hash_reverse_apply (ht=ht@entry=0xdba9f0 <executor_globals+304>,
    apply_func=apply_func@entry=0x5ef8e0 <zval_call_destructor>)
    at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:1602
#12 0x00000000005efd1d in shutdown_destructors ()
    at /root/basesys-components/php-7.1.25/Zend/zend_execute_API.c:242
#13 0x00000000006003f4 in zend_call_destructors ()
    at /root/basesys-components/php-7.1.25/Zend/zend.c:990
#14 0x00000000005a0275 in php_request_shutdown (dummy=dummy@entry=0x0)
    at /root/basesys-components/php-7.1.25/main/main.c:1849
#15 0x0000000000699e4b in do_cli (argc=4, argv=0xdbfbe0)
    at /root/basesys-components/php-7.1.25/sapi/cli/php_cli.c:1160
#16 0x0000000000427800 in main (argc=4, argv=0xdbfbe0)
    at /root/basesys-components/php-7.1.25/sapi/cli/php_cli.c:1381

**Comment by vityank** [comment]
I've made a debug version of my PHP with STREAM_DEBUG also enabled.
With PHP/debug is fails slightly before the release version, but as i see the release of second Redis object which instantiated in a shutdown function(@ 0x107cf00) causes the previous connection object (@ 0x107cf90) to become corrupted(Like in use, after free scenarios): ops = 0x1, ctx = 0x100007f00000002 - both are obviously invalid(Non-debug crashes accessing ops, debug crashes accessing ctx, before ops).

Starting program: /opt/php71/bin/php /root/phpredis/testCloseOnShutdown.php
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
stream_alloc: STDIO:0x7ffff3260280 persistent=(null)
stream_alloc: STDIO:0x7ffff3260500 persistent=(null)
stream_alloc: STDIO:0x7ffff3260640 persistent=(null)
Start test
stream_alloc: tcp_socket/ssl:0x107cf90 persistent=phpredis:localhost:6379:
Res: OK
stream_free: tcp_socket/ssl:0x107cf90[(null)] in_free=0 opts=CALL_DTOR, RELEASE_STREAM, PERSISTENT
stream_free: tcp_socket/ssl:0x107cf90[(null)] preserve_handle=0 release_cast=1 remove_rsrc=1
stream_free: tcp_socket/ssl:0x107cf90[(null)] in_free=1 opts=CALL_DTOR, RELEASE_STREAM, RSRC_DTOR
Finishing test
stream_alloc: tcp_socket/ssl:0x107cf00 persistent=phpredis:localhost:6379:
Res: OK
stream_free: tcp_socket/ssl:0x107cf00[(null)] in_free=0 opts=CALL_DTOR, RELEASE_STREAM, PERSISTENT
stream_free: tcp_socket/ssl:0x107cf00[(null)] preserve_handle=0 release_cast=1 remove_rsrc=1
stream_free: tcp_socket/ssl:0x107cf00[(null)] in_free=1 opts=CALL_DTOR, RELEASE_STREAM, RSRC_DTOR

Program received signal SIGSEGV, Segmentation fault.
0x00000000006a666a in _php_stream_free (stream=0x107cf90, close_options=19)
    at /root/basesys-components/php-7.1.25/main/streams/streams.c:372
372                     context = PHP_STREAM_CONTEXT(stream);


(gdb) bt
#0  0x00000000006a666a in _php_stream_free (stream=0x107cf90, close_options=19)
    at /root/basesys-components/php-7.1.25/main/streams/streams.c:372
#1  0x00007fffe4545b90 in redis_sock_disconnect (redis_sock=0x7ffff3273460, force=1)
    at /root/phpredis/library.c:1865
#2  0x00007fffe45063c2 in zim_Redis_close (execute_data=0x7ffff32150c0, return_value=0x7fffffffc6a0)
    at /root/phpredis/redis.c:1015
#3  0x000000000077b822 in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER ()
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:970
#4  0x000000000077a218 in execute_ex (ex=0x7ffff3215030)
    at /root/basesys-components/php-7.1.25/Zend/zend_vm_execute.h:429
#5  0x0000000000701dfd in zend_call_function (fci=0x7fffffffc8d0, fci_cache=0x7fffffffc870)
    at /root/basesys-components/php-7.1.25/Zend/zend_execute_API.c:855
#6  0x0000000000742382 in zend_call_method (object=0x7fffffffc9e0, obj_ce=0x7ffff3204018, fn_proxy=0x7fffffffc9f0,
    function_name=0xc22e8e "__destruct", function_name_len=10, retval_ptr=0x0, param_count=0, arg1=0x0, arg2=0x0)
    at /root/basesys-components/php-7.1.25/Zend/zend_interfaces.c:99
#7  0x0000000000763389 in zend_objects_destroy_object (object=0x7ffff325a720)
    at /root/basesys-components/php-7.1.25/Zend/zend_objects.c:156
#8  0x0000000000769f83 in zend_objects_store_del (object=0x7ffff325a720)
    at /root/basesys-components/php-7.1.25/Zend/zend_objects_API.c:162
#9  0x0000000000717656 in _zval_dtor_func (p=0x7ffff325a720,
    __zend_filename=0xc1cab0 "/root/basesys-components/php-7.1.25/Zend/zend_variables.c", __zend_lineno=205)
    at /root/basesys-components/php-7.1.25/Zend/zend_variables.c:56
#10 0x0000000000717536 in i_zval_ptr_dtor (zval_ptr=0x7fffffffcc20,
    __zend_filename=0xc1cab0 "/root/basesys-components/php-7.1.25/Zend/zend_variables.c", __zend_lineno=205)
    at /root/basesys-components/php-7.1.25/Zend/zend_variables.h:48
#11 0x0000000000717c99 in _zval_ptr_dtor_wrapper (zval_ptr=0x7fffffffcc20)
    at /root/basesys-components/php-7.1.25/Zend/zend_variables.c:205
#12 0x000000000072f02f in _zend_hash_del_el_ex (ht=0xf89210 <executor_globals+304>, idx=7, p=0x7ffff325b1e0,
    prev=0x0) at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:997
#13 0x000000000072f0fa in _zend_hash_del_el (ht=0xf89210 <executor_globals+304>, idx=7, p=0x7ffff325b1e0)
    at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:1020
#14 0x0000000000730984 in zend_hash_reverse_apply (ht=0xf89210 <executor_globals+304>,
    apply_func=0x6ffb41 <zval_call_destructor>) at /root/basesys-components/php-7.1.25/Zend/zend_hash.c:1602
#15 0x00000000006ffd9f in shutdown_destructors ()
    at /root/basesys-components/php-7.1.25/Zend/zend_execute_API.c:242
#16 0x0000000000719c03 in zend_call_destructors () at /root/basesys-components/php-7.1.25/Zend/zend.c:990
#17 0x00000000006888f5 in php_request_shutdown (dummy=0x0) at /root/basesys-components/php-7.1.25/main/main.c:1849
#18 0x00000000007fbc23 in do_cli (argc=2, argv=0xf8ebd0)
    at /root/basesys-components/php-7.1.25/sapi/cli/php_cli.c:1160
#19 0x00000000007fc307 in main (argc=2, argv=0xf8ebd0)
    at /root/basesys-components/php-7.1.25/sapi/cli/php_cli.c:1381


(gdb) print *stream
$1 = {ops = 0x1, abstract = 0x0, readfilters = {head = 0x0, tail = 0x0, stream = 0x2a}, writefilters = {
    head = 0x0, tail = 0x2000, stream = 0x25}, wrapper = 0x25, wrapperthis = 0x2000, wrapperdata = {value = {
      lval = 140737024354017, dval = 6.9533328831242653e-310, counted = 0x7fffe457e6e1, str = 0x7fffe457e6e1,
      arr = 0x7fffe457e6e1, obj = 0x7fffe457e6e1, res = 0x7fffe457e6e1, ref = 0x7fffe457e6e1,
      ast = 0x7fffe457e6e1, zv = 0x7fffe457e6e1, ptr = 0x7fffe457e6e1, ce = 0x7fffe457e6e1, func = 0x7fffe457e6e1,
      ww = {w1 = 3830965985, w2 = 32767}}, u1 = {v = {type = 1 '\001', type_flags = 7 '\a',
        const_flags = 0 '\000', reserved = 0 '\000'}, type_info = 1793}, u2 = {next = 0, cache_slot = 0,
      lineno = 0, num_args = 0, fe_pos = 0, fe_iter_idx = 0, access_flags = 0, property_guard = 0, extra = 0}},
  fgetss_state = 0, is_persistent = 0, mode = "\221\000\000\000\000\000\000\000\070\bN\365\377\177\000",
  res = 0x7ffff54e0838 <main_arena+216>, in_free = 16, fclose_stdiocast = 0, stdiocast = 0x107d030, __exposed = 0,
  orig_path = 0x107e3e0 "\b\nN\365\377\177", ctx = 0x100007f00000002, flags = 0, eof = 0, position = 42,
  readbuf = 0x41 <Address 0x41 out of bounds>, readbuflen = 140737308919784, readpos = 140737308919784,
  writepos = 37, chunk_size = 8192, open_filename = 0x7fffe457e6e1 "/root/phpredis/library.c", open_lineno = 1793,
  enclosing_stream = 0x190}

**Comment by michael-grunder** [comment]
Yeah, this bug is just _weird_.  

For example, if I change the reproduction script [to this](https://gist.github.com/michael-grunder/22c25def77fbf05fc0ab87d620e777a1) then I get a segfault if I don't pass any arguments but not if I do.

The only change is that if passed arguments, the `testSet` function uses `uniqid` instead of `microtime(true)` to generate the key value.

Furthermore, in my VM I can cause or _not_ cause a segault based on _where_ I put the script:

```bash
[root@b02dbd7a1db1 phpredis]# php -d extension=modules/redis.so segfault.orig.php
Start test
Res: OK
Finishing test
Res: OK
[root@b02dbd7a1db1 phpredis]# php -d extension=modules/redis.so /root/segfault.orig.php
Start test
Res: OK
Finishing test
Res: OK
PHP Warning:  Unknown list entry type (0) in /root/segfault.orig.php on line 30
Segmentation fault (core dumped)
[root@b02dbd7a1db1 phpredis]# diff segfault.orig.php /root/segfault.orig.php
[root@b02dbd7a1db1 phpredis]#
```

Continuing to try to pin this down.

**Comment by yatsukhnenko** [comment]
Heisenbug :smile:

**Comment by vityank** [comment]
I think, I realized what's going on, but i may be wrong.

(a) Global variable get stream allocated at address p1
(b1) Local variable get stream allocated at address p1 <- persistent list
(b2) Local variable deallocated, _php_stream_free(PERS=1) removes the p1 from persistent_list
* At this point p1 is dangling pointer
(c) We allocate new local variable in shutdown function:
   (c1) We are lucky and got... p1 again, newly allocated and added to persistent_list
          We perform two free's on p1 in sequence w/o something in the middle writing to it.
          Between the two frees our p1 becomes dangling as well but we are lucky with the timing,
                 bug still in place, just gets annoticed.
           Our script finishes fine.
--- OR ---
   (c2) We get new handle at p2... and p2 is slightly lower than our dangling p1...
          PHP fills the php_stream at p2, overwriting part of memory region held by dangling p1...
          We free the p2 - OK.
          We access the p1 and - Segfault :)

Looks like reference counting shall be implemented or ref-counter checked then calling php_stream_pclose...   Or maybe attempt to call php_stream_close for persistent too to avoid removing the stream from persistent_list - calls to php_stream_alloc will reutilize the object from plist again...

**Comment by vityank** [comment]
Thanks @michael-grunder . Happy holidays and thank you for a hard work on this project :)

**Comment by yatsukhnenko** [comment]
```C
diff --git a/library.c b/library.c
index d3278a8..9b7372d 100644
--- a/library.c
+++ b/library.c
@@ -1866,7 +1866,7 @@ redis_sock_disconnect(RedisSock *redis_sock, int force TSRMLS_DC)
         return FAILURE;
     } else if (redis_sock->stream) {
         if (redis_sock->persistent) {
-            if (force) {
+            if (force && !redis_sock->stream->in_free) {
                 php_stream_pclose(redis_sock->stream);
             }
         } else {
```

**Comment by michael-grunder** [comment]
Hi @vityank,

I'll take a closer look at that patch in the morning.  The issue is a bit tricky because what we really want to do is close the underlying socket but keep the stream in a safe way.  The reason we switched to `redis_stream_pclose` here is because people were having issues with persistent connections getting out of sync after a network error.

There does appear to be a clean way to accomplish this in PHP >= 7.0:

```diff --git a/library.c b/library.c
index d3278a8..f10e1f5 100644
--- a/library.c
+++ b/library.c
@@ -1856,6 +1856,14 @@ redis_sock_server_open(RedisSock *redis_sock TSRMLS_DC)
     return res;
 }

+/* Disconnect the underlying socket without freeing the php_stream itself. This
+ * lets us reset a persistent connection that has gotten itself into a bad state
+ * without risking memory errors as multiple Redis objects can point to the same
+ * persistent stream. */
+static void redis_stream_pclose(php_stream *stream) {
+    php_stream_free(stream, PHP_STREAM_FREE_KEEP_RSRC | PHP_STREAM_FREE_CALL_DTOR);
+}
+
 /**
  * redis_sock_disconnect
  */
@@ -1864,16 +1872,18 @@ redis_sock_disconnect(RedisSock *redis_sock, int force TSRMLS_DC)
 {
     if (redis_sock == NULL) {
         return FAILURE;
-    } else if (redis_sock->stream) {
-        if (redis_sock->persistent) {
-            if (force) {
-                php_stream_pclose(redis_sock->stream);
-            }
-        } else {
+    }
+
+    if (redis_sock->stream) {
+        if (!redis_sock->persistent) {
             php_stream_close(redis_sock->stream);
+        } else if (force) {
+            redis_stream_pclose(redis_sock->stream);
         }
+
         redis_sock->stream = NULL;
     }
+
     redis_sock->mode = ATOMIC;
     redis_sock->status = REDIS_SOCK_STATUS_DISCONNECTED;
     redis_sock->watching = 0;
```

Unfortunately `PHP_STREAM_FREE_KEEP_RSRC` doesn't seem to exist in PHP 5.x so I need to determine what an alternative approach might be for older versions of PHP.

**Comment by o5** [comment]
@michael-grunder 

> Unfortunately PHP_STREAM_FREE_KEEP_RSRC doesn't seem to exist in PHP 5.x so I need to determine what an alternative approach might be for older versions of PHP.

You're awesome OSS maintainer, thanks! I don't understand completely this PHP extension internals but PHP 5.x will be end of life [in next week](http://php.net/supported-versions.php). What about drop supporting this old version?

I'm going to upgrade Redis instances to latest version 5.0.2 in our projects but I don't know if I understand this issue correctly: Right now, I use `phpredis/phpredis@3.1.6` with  `RedisCluster` without persistent connection. This issue will not be in my use-case with `phpredis/phpredis@4.2.0`, right?

Thanks again.

**Comment by vityank** [comment]
I know what the major change is to support reconnects in case of problems(Then built-in keep-alive in PHP doesn't fit, probably due to unlimited timeouts). However, @michael-grunder , I afraid approach with PHP_STREAM_FREE_KEEP_RSRC/PHP_STREAM_FREE_PRESERVE_HANDLE will not work, because bug has another background(I hope i wrong, ofc). Somehow we get PHP to store our stream in both persistent and regular lists, while refcounting increases only on regular list, while php_stream_pclose goes to persistent list directly, and this one has refcnt=1 because of this design decision by PHP team.

The bug's origin is in php_stream_from_persistent_id:

``
PHPAPI int php_stream_from_persistent_id(const char *persistent_id, php_stream **stream)
{
    zend_resource *le;

    if ((le = zend_hash_str_find_ptr(&EG(persistent_list), persistent_id, strlen(persistent_id))) != NULL) {
        if (le->type == le_pstream) {
            if (stream) {
                zend_resource *regentry = NULL;

                /* see if this persistent resource already has been loaded to the
                 * regular list; allowing the same resource in several entries in the
                 * regular list causes trouble (see bug #54623) */
                *stream = (php_stream*)le->ptr;
                ZEND_HASH_FOREACH_PTR(&EG(regular_list), regentry) {
                    if (regentry->ptr == le->ptr) {
                        GC_REFCOUNT(regentry)++;
                        (*stream)->res = regentry;
                        return PHP_STREAM_PERSISTENT_SUCCESS; <===== rfc for le not increased here, only for regentry 'clone'...
                    }
                } ZEND_HASH_FOREACH_END();
                GC_REFCOUNT(le)++;
                (*stream)->res = zend_register_resource(*stream, le_pstream);
            }
            return PHP_STREAM_PERSISTENT_SUCCESS;
        }
        return PHP_STREAM_PERSISTENT_FAILURE;
    }
    return PHP_STREAM_PERSISTENT_NOT_EXIST;
}
``

**Comment by michael-grunder** [comment]
Hey @vityank,

> In any case closing one of persistent connections to same host will close all others as well(And developers probably don't want this to happen).

Yeah, I think we're running into conflicting requirements here.  On the one hand, we recently changed the way `close` works in order to fix the problem of out-of-sync persistent connections, but as it sits now this is just broken (segfault/invalid reads).  It's probably silently working for many users because they only have one object to any given `{host}:{port}` and/or aren't calling `close()` manually.

I'm going to be honest here and admit that until recently, I was mistaken about how PHP's persistent connections worked.  I've never used them in production, and have only ever played with them in order to fix various isues here.  

I was surprised by what this script does:
```php
<?php

$obj_r1 = new Redis();
$obj_r2 = new Redis();

$obj_r1->pconnect('localhost', 6379);
$obj_r2->pconnect('localhost', 6379);

$obj_r1->multi();

// WHAT?!: This will print: +QUEUED
echo $obj_r2->ping('dummy') . "\n";
?>
```

I assumed PHP used some sort of "in use" / "free" mechanism with persistent connections, but it appears to be a shared free-for-all.

So I think that the real solution to this is going to need to be more substantial.  We may need to revert part of 1d997873 that invokes `php_stream_pclose` on normal calls to `Redis->close()`, and perhaps create new method `Redis->pclose()` and document that this will close connections to all objects sharing the same connection, and even if we do that this isn't going to work for PHP 5 (maybe we can use this as an excuse to drop PHP5 support!).

I didn't do extensive testing, but for PHP >= 7.0 objects sharing the same persistent connection do continue to operate because they will automatically reconnect on the next command.  It will of course break any shared connections inside of a `multi()...exec()` but that feels broken to me as it is.

@yatsukhnenko Curious as to any thoughts you have about this issue :smile:

With all of that being said, I think the _real_ solution here is to actually implement persistent connections in a sane way, by implementing the concept of `free` or `in use`, which appears to be what [mysqli does](https://github.com/php/php-src/blob/731b82eb715b4076499d0a1c599de1a7ef98992d/ext/mysqli/mysqli_nonapi.c#L164)

---

@o5: You shouldn't be affected if you aren't using persistent connections.  I'd still recommend testing before deploying in production just to be safe. :smile:

TL;DR:  This is going to require something substantial in order to actually fix the issue in a sane way.

**Comment by vityank** [comment]
> I assumed PHP used some sort of "in use" / "free" mechanism with persistent connections, but it appears to be a shared free-for-all.

Many extensions and PHP's internal code uses this approach :) It can be controlled in some of them then C developer exposes extra argument for custom persistent id's - using different ones it really creates different persistent elements. However many devs are hiding them by using host+port+optionally some id+extra connection param such as timeout to generate those, and you do the same. It works as it should unless no manual pclose calls are used. PHP stores single object under the hood for each host:port pair and it is shared by the multiple objects - expected to never be closed by external code and live with fcgi child/FPM child process during all it's lifetime, surviving hundreds of requests passing and allowing high QPS rates. This by design will not work with our demos($obj_r1,$obj_r2)  but perfectly fits the Singletone design $r = redisCluster->getConnection() :)
Zend devs included simple keep-alive in the transports API to reconnect in case of problems, but they force unlimited(0) timeout for this operation w/o allowing dev to pass it's own value. In a case of host, which got down a call will be blocked for socket timeout configured in the OS which is probably 30/60 secs and is very high. Here it splits into target use cases: (1) Redis server is one of main components and if it n/a then net i/o blocked software not pose the problem because main part of the system already malfunctioned. (2) General multipurpose code working with multiple servers - This software may not allow long blocks, according to project requirements and targets.

> So I think that the real solution to this is going to need to be more substantial. We may need to revert part of 1d99787 that invokes php_stream_pclose on normal calls to Redis->close(), and perhaps create new method Redis->pclose()

Looks like sane and also backward compatible solution. It can also possibly be supported for PHP 5 as well if phpredis will store hashmap with all its persistent connections and it's close/reopen cycle will affect all targeted Redis objects using one write to the hashmap. It will however require replacing all redisObj->stream in C code with something like PHPREDIS_GETSTREAM(redisObj) with logic of direct retrieval from object or reading from hashmap inside function/macro. The problem is that we need a 'reopen' action - not just disconnect, while close() will not do anything for persistent ones - They are there to stay, by the design and main use-cases of persistent connections - always have socket open to specified host:port...

**Comment by michael-grunder** [comment]
OK, so I think in the short term I'll update the code to revert `close()` to previous functionality, and add `pclose()` _at least_ for PHP >= 7.0 since it can be done with very little modification to the code.

> The problem is that we need a 'reopen' action - not just disconnect

From what I've read, in addition to implementing a proper connection pool the `mysqli` extension also attempts to clean up connections in an invalid state by calling `mysql_change_user()`.  Theoretically, we might be able to get Redis to help us here with something like: `CLIENT KILL ID [our id] SKIPME NO` and then rely on the existing automatic reconnect logic, but this is in no way a thought through idea. :smile:

**Comment by yatsukhnenko** [comment]
I'm currently on vacation  :snowman: so I can't dig into this issue to much

**Comment by yatsukhnenko** [comment]
BTW I like an idea about connection pool and we have another issues where users ask about this

**Comment by yatsukhnenko** [comment]
@michael-grunder Let's add object handle id to all persistent connections?
```diff --git a/common.h b/common.h
index adbe303..02ae95e 100644
--- a/common.h
+++ b/common.h
@@ -684,6 +684,7 @@ typedef struct {
 
     int            readonly;
     int            tcp_keepalive;
+    uint32_t       handle;
 } RedisSock;
 /* }}} */
 
diff --git a/library.c b/library.c
index d3278a8..c700602 100644
--- a/library.c
+++ b/library.c
@@ -1785,11 +1785,11 @@ PHP_REDIS_API int redis_sock_connect(RedisSock *redis_sock TSRMLS_DC)
 
     if (redis_sock->persistent) {
         if (redis_sock->persistent_id) {
-            spprintf(&persistent_id, 0, "phpredis:%s:%s", host,
-                ZSTR_VAL(redis_sock->persistent_id));
+            spprintf(&persistent_id, 0, "phpredis:%s:%s:%u", host,
+                ZSTR_VAL(redis_sock->persistent_id), redis_sock->handle);
         } else {
-            spprintf(&persistent_id, 0, "phpredis:%s:%f", host,
-                redis_sock->timeout);
+            spprintf(&persistent_id, 0, "phpredis:%s:%f:%u", host,
+                redis_sock->timeout, redis_sock->handle);
         }
     }
 
diff --git a/redis.c b/redis.c
index 6f712d7..de6e235 100644
--- a/redis.c
+++ b/redis.c
@@ -629,6 +629,7 @@ redis_sock_get_instance(zval *id TSRMLS_DC, int no_throw)
     if (Z_TYPE_P(id) == IS_OBJECT) {
         redis = PHPREDIS_GET_OBJECT(redis_object, id);
         if (redis->sock) {
+            redis->sock->handle = Z_OBJ_HANDLE_P(id);
             return redis->sock;
         }
     }
@@ -972,6 +973,7 @@ redis_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent)
     redis->sock = redis_sock_create(host, host_len, port, timeout, read_timeout, persistent,
         persistent_id, retry_interval);
 
+    redis->sock->handle = Z_OBJ_HANDLE_P(object);
     if (redis_sock_server_open(redis->sock TSRMLS_CC) < 0) {
         if (redis->sock->err) {
             zend_throw_exception(redis_exception_ce, ZSTR_VAL(redis->sock->err), 0 TSRMLS_CC);
```

**Comment by michael-grunder** [comment]
@yatsukhnenko Why are you working on your vacation!  :joy: 

This does solve the problem but I worry that it would massively increase file descriptor pressure.  I think this is going to require a proper solution (persistent connection pooling).

**Comment by yatsukhnenko** [comment]
@vityank could you test changes from `connection-pool` branch?

**Comment by vityank** [comment]
@yatsukhnenko , Seem to work stable with redis.pconnect.pooling_enabled=1(Didn't checked for memleaks or stability over long time during high load),
Has different random crashes w/o the new option.

**Comment by yatsukhnenko** [comment]
I merged connections pooling functionality into `develop` branch so let's say that this issue is fixed :smile: @vityank, @michael-grunder Feel free to reopen this issue if it doesn't fixed on your end.
